framework/include/systems/NonlinearSystemBase.h

```diff
@@ -227,6 +227,45 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
    */
   std::shared_ptr<Split> getSplit(const std::string & name);
 
+  /**
+   * We offer the option to check convergence against the pre-SMO residual. This method handles the
+   * logic as to whether we should perform such residual evaluation.
+   *
+   * @return A boolean indicating whether we should evaluate the pre-SMO residual.
+   */
+  bool shouldEvaluatePreSMOResidual() const;
+
+  /**
+   * Returns a writable reference to a boolean controlling whether to evaluate the pre-SMO residual
+   * and use it in the subsequent relative convergence checks.
+   *
+   * If set to true, an _additional_ residual evaluation is performed before any
+   * solution-modifying object is executed, and before the initial (0-th nonlinear iteration)
+   * residual evaluation. Such residual is referred to as the pre-SMO residual. If the pre-SMO
+   * residual is evaluated, it is used in the subsequent relative convergence checks.
+   *
+   * If set to false, no residual evaluation takes place before the initial residual evaluation, and
+   * the initial residual is used in the subsequent relative convergence checks. This mode is
+   * recommended for performance-critical code as it avoids the additional pre-SMO residual
+   * evaluation.
+   */
+  bool & usePreSMOResidual() { return _use_pre_smo_residual; }
```

Comment:
can we change this to accept an argument to perform the write that way instead of releasing the data member as a writable reference? This allows some degree of encapsulation such that if we need to add some sanity checks (or other things) in the future, we can do so. We  have discovered the hard way in libMesh that by releasing writable references to data members, we have no way of tracking whether a class  has been modified or not

Reply:
> we have no way of tracking whether a class has been modified or not

well you could return a wrapper around boolean e.g. `Ref<bool>` that overloads `operator=`, but I see your point.